packages/playwright/src/matchers/toMatchText.ts

```diff
@@ -78,42 +79,80 @@ export async function toMatchText(
         printedReceived = `Received string: ${formattedReceived}`;
       }
     } else {
       if (notFound) {
-        printedExpected = `Expected pattern: not ${this.utils.printExpected(expected)}`;
+        printedExpected = `Expected pattern: not ${state.utils.printExpected(expected)}`;
         printedReceived = `Received: ${received}`;
       } else {
-        printedExpected = `Expected pattern: not ${this.utils.printExpected(expected)}`;
+        printedExpected = `Expected pattern: not ${state.utils.printExpected(expected)}`;
         const formattedReceived = printReceivedStringContainExpectedResult(receivedString, typeof expected.exec === 'function' ? expected.exec(receivedString) : null);
         printedReceived = `Received string: ${formattedReceived}`;
       }
     }
   } else {
-    const labelExpected = `Expected ${typeof expected === 'string' ? stringSubstring : 'pattern'}`;
+    const labelExpected = `Expected ${typeof expected === 'string' ? stringName : 'pattern'}`;
     if (notFound) {
-      printedExpected = `${labelExpected}: ${this.utils.printExpected(expected)}`;
+      printedExpected = `${labelExpected}: ${state.utils.printExpected(expected)}`;
       printedReceived = `Received: ${received}`;
     } else {
-      printedDiff = this.utils.printDiffOrStringify(expected, receivedString, labelExpected, 'Received string', false);
+      printedDiff = state.utils.printDiffOrStringify(expected, receivedString, labelExpected, 'Received string', false);
     }
   }
 
-  const message = () => {
-    const resultDetails = printedDiff ? printedDiff : printedExpected + '\n' + printedReceived;
-    return messagePrefix + resultDetails + callLogText(log);
-  };
+  const resultDetails = printedDiff ? printedDiff : printedExpected + '\n' + printedReceived;
+  return messagePrefix + resultDetails + callLogText(callLog);
+}
+
+export async function toMatchText(
+  this: ExpectMatcherState,
+  matcherName: string,
+  receiver: Locator,
+  receiverType: string,
+  query: (isNot: boolean, timeout: number) => Promise<{ matches: boolean, received?: string, log?: string[], timedOut?: boolean }>,
+  expected: string | RegExp,
+  options: { timeout?: number, matchSubstring?: boolean } = {},
+): Promise<MatcherResult<string | RegExp, string>> {
+  expectTypes(receiver, [receiverType], matcherName);
+  toMatchExpectedStringVerification(this, matcherName, receiver, receiver, expected);
+
+  const timeout = options.timeout ?? this.timeout;
+
+  const { matches: pass, received, log, timedOut } = await query(!!this.isNot, timeout);
+  if (pass === !this.isNot) {
+    return {
+      name: matcherName,
+      message: () => '',
+      pass,
+      expected
+    };
+  }
+
+  const stringSubstring = options.matchSubstring ? 'substring' : 'string';
 
   return {
     name: matcherName,
     expected,
-    message,
+    message: () =>
```

Comment:
The way I see it you are paying 14 lines of a call site + 13 lines of a declaration for 40 lines of reusable code that will only hold you back because toMatchText and toHaveURL are going to be conceptually different and will want different treatment. If that's the only reason you are extracting this code - don't. Copy-paste of highly customizable code is always better long-term. It keeps things simple, it allows for better ergonomics, it does not make you cut corners (hard-coded expected, pre-computed message()).